doc: Add custom runtime with PVC example (#495)
kserve/modelmesh-serving: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md

# The Python-Based Custom Runtime with Model Stored on Persistent Volume Claim

> This document can provide the step by step instructions to demonstrate how to write a custom python-based `ServingRuntime` of `MLModel` and store the trained models on persistent volume claims.

Reviewer: ```suggestion
This document provides step-by-step instructions to demonstrate how to write a custom Python-based `ServingRuntime` inheriting from [MLServer's MLModel class](https://github.com/SeldonIO/MLServer/blob/master/mlserver/model.py) and deploy a model stored on persistent volume claims with it. 
```

Author: Done!

> Before writing the custom `ServingRuntime`, user should have a K8s cluster with `kserve modelmesh` installed firstly by following [doc](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md). Please make sure `modelMesh serving` is working well, it looks like below:
> 
> The `modelMesh serving` pods are running:
> 
> ```shell
> kubectl get pods
> 
> NAME                                    READY   STATUS    RESTARTS   AGE
> etcd-78ff7867d5-h4r2t                   1/1     Running   0          5m
> minio-6b85d9b464-nmk5b                  1/1     Running   0          5m
> modelmesh-controller-695b8488f5-hndgn   1/1     Running   0          5m
> ```
> 
> The `ServingRuntime`s are available: 
> 
> ```shell
> kubectl get servingruntimes
> 
> NAME             DISABLED   MODELTYPE     CONTAINERS   AGE
> mlserver-1.x                sklearn       mlserver     2m
> ovms-1.x                    openvino_ir   ovms         2m
> torchserve-0.x              pytorch-mar   torchserve   2m
> triton-2.x                  keras         triton       2m
> ```

Reviewer: ```suggestion
This example assumes that ModelMesh Serving was deployed using the [quickstart guide](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md). 
```

Author: Done!

# Deploy a model stored on a Persistent Volume Claim

Let's use namespace `modelmesh-serving` here:

```shell
kubectl config set-context --current --namespace=modelmesh-serving
```

## 1. Create PV and PVC for storing model file

```shell
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: PersistentVolume
metadata:

>   name: my-models-pv

Reviewer: ```suggestion
  name: my-models-pvc
```

Author: Hi @rafvasq, this yaml file firstly create persistent volume, then persistent volume claim, so I use the different name for it.

      persistentVolumeClaim:
        claimName: "my-models-pvc"
EOF
```

## 3. Store the model on this persistent volume

The sample model file we used in this doc is `sklearn/mnist-svm.joblib`.

```shell
curl -sOL https://github.com/kserve/modelmesh-minio-examples/raw/main/sklearn/mnist-svm.joblib
```

Copy this model file to the `pvc-access` pod:

```shell
kubectl cp mnist-svm.joblib pvc-access:/mnt/models/
```

Verify the model exists on the persistent volume：

```shell
kubectl exec -it pvc-access -- ls -alr /mnt/models/

# total 348
# -rw-rw-r-- 1 1000 1000 344817 Mar 19 08:37 mnist-svm.joblib
# drwxr-xr-x 1 root root   4096 Mar 19 08:34 ..
# drwxr-xr-x 2 root root   4096 Mar 19 08:37 .
```

## 4. Configure ModelMesh Serving to use the persistent volume claim

Create the `model-serving-config` ConfigMap with the setting allowAnyPVC: true:

```shell
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: model-serving-config
data:
  config.yaml: |
    allowAnyPVC: true
EOF
```

Verify the configuration setting:

```shell
kubectl get cm "model-serving-config" -o jsonpath="{.data['config\.yaml']}"
```

# Implement the Python-based Custom Runtime on MLServer

> All necessary resources are in [custom-model](./custom-model), including source code file and Dockerfile.

Reviewer: ```suggestion
All of the necessary resources are contained in [custom-model](./custom-model), including the model code, and the Dockerfile.
```

Author: Done.

## 1. Implement the API of MLModel

> Please understand the `Custom MLModel Template` before writing the real code by following [Python-based Custom Runtime on MLServer](../mlserver_custom.md). Both `load` and `predict` must be implemented to support the custom `ServingRuntime`, and below snippet can be the simplified implementation of `CustomMLModel` for model `mnist-svm.joblib`.

Reviewer: ```suggestion
Both `load` and `predict` must be implemented to support this custom `ServingRuntime`. The snippet below provides a simplified implementation of `CustomMLModel` for model `mnist-svm.joblib`. You can read more about it [here](https://github.com/kserve/modelmesh-serving/blob/main/docs/runtimes/mlserver_custom.md).
```

Author: Done

            self._model = load(model_uri)
            logger.info("Model loaded successfully")
        else:
            logger.info(f"Model not exist in {model_uri}")
            # raise FileNotFoundError(model_uri)
            self.ready = False
            return self.ready
    
        self.ready = True
        return self.ready

    async def predict(self, payload: InferenceRequest) -> InferenceResponse:
        input_data = [input_data.data for input_data in payload.inputs]
        input_name = [input_data.name for input_data in payload.inputs]
        input_data_array = np.array(input_data)
        result = self._model.predict(input_data_array) 
        predictions = np.array(result)

        logger.info(f"Predict result is: {result}")
        return InferenceResponse(
            id=payload.id,
            model_name = self.name,
            model_version = self.version,
            outputs = [
                ResponseOutput(
                    name = str(input_name[0]),
                    shape = predictions.shape,
                    datatype = "INT64",
                    data=predictions.tolist(),
                )
            ],
        )

> ```

Reviewer: I think you can just link to the `custom-model.py` here.

Author: Yeah, make sense, I link to `custom-model.py`

> ## 2. Build custom ServingRuntim image

Reviewer: ```suggestion
## 2. Build the custom `ServingRuntime` image
```

Author: Done

> User can use `mlserver` or `docker` to help to build the custom `ServingRuntim` image, and the latter is chosen in this doc. The available Dockerfile is below:

Reviewer: ```suggestion
You can use [`mlserver`](https://mlserver.readthedocs.io/en/stable/examples/custom/README.html#building-a-custom-image) or `docker` to help to build the custom `ServingRuntime` image, and the latter is done below. 
```

Author: Done

# single-file module, this can be accomplished with:
COPY --chown=${USER} ./custom_model.py /opt/custom_model.py
ENV PYTHONPATH=/opt/
WORKDIR /opt

# environment variables to be compatible with ModelMesh Serving
#  these can also be set in the ServingRuntime, but this is recommended for
#  consistency when building and testing
# reference: https://mlserver.readthedocs.io/en/latest/reference/settings.html
ENV MLSERVER_MODELS_DIR=/models/_mlserver_models \
    MLSERVER_GRPC_PORT=8001 \
    MLSERVER_HTTP_PORT=8002 \
    MLSERVER_METRICS_PORT=8082 \
    MLSERVER_LOAD_MODELS_AT_STARTUP=false \
    MLSERVER_DEBUG=false \
    MLSERVER_PARALLEL_WORKERS=1 \
    MLSERVER_GRPC_MAX_MESSAGE_LENGTH=33554432 \
    # https://github.com/SeldonIO/MLServer/pull/748
    MLSERVER__CUSTOM_GRPC_SERVER_SETTINGS='{"grpc.max_metadata_size": "32768"}' \
    MLSERVER_MODEL_NAME=dummy-model

# With this setting, the implementation field is not required in the model
# settings which eases integration by allowing the built-in adapter to generate
# a basic model settings file
ENV MLSERVER_MODEL_IMPLEMENTATION=custom_model.CustomMLModel

CMD mlserver start $MLSERVER_MODELS_DIR

> ```

Reviewer: Similar to the model above, you can just link to the `Dockerfile`

Author: Yeah, make sense, link to `custom_mode/Dockerfile`

> Get into the folder [custom-model](./custom-model), then execute command:

Reviewer: ```suggestion
To build the image, execute the following command from within the [custom-model](./custom-model) directory.
```

Author: Done

```shell
docker build -t <DOCKER-HUB-ORG>/custom-model-server:0.1 .

```

> **Note**: Please use the `--build-arg` to add the http proxy if there is proxy in user's environment, such as: 
```shell
docker build --build-arg HTTP_PROXY=http://<DOMAIN-OR-IP>:PORT --build-arg HTTPS_PROXY=http://<DOMAIN-OR-IP>:PORT -t <DOCKER-HUB-ORG>/custom-model-server:0.1 .
```

## 3. Define and Apply Custom ServingRuntime

> Please understand the `Custom ServingRuntime Template` before defining it by following [Python-based Custom Runtime on MLServer](../mlserver_custom.md). Executing below command to create a custom ServingRuntime.

Reviewer: ```suggestion
Below, you will create a ServingRuntime using the image built above. You can learn more about the custom `ServingRuntime` template [here](https://github.com/kserve/modelmesh-serving/blob/main/docs/runtimes/mlserver_custom.md#custom-servingruntime-template).
```

Author: Done

        - name: MLSERVER_GRPC_PORT
          value: "8001"
        - name: MLSERVER_HTTP_PORT
          value: "8002"
        - name: MLSERVER_LOAD_MODELS_AT_STARTUP
          value: "false"
        - name: MLSERVER_MODEL_NAME
          value: dummy-model
        - name: MLSERVER_HOST
          value: "127.0.0.1"
        - name: MLSERVER_GRPC_MAX_MESSAGE_LENGTH
          value: "-1"
      resources:
        requests:
          cpu: 500m
          memory: 1Gi
        limits:
          cpu: "5"
          memory: 1Gi
  builtInAdapter:
    serverType: mlserver
    runtimeManagementPort: 8001
    memBufferBytes: 134217728
    modelLoadingTimeoutMillis: 90000
EOF
```

Verify the available `ServingRuntime`, including the custom one: 

```shell
kubectl get servingruntimes

NAME                  DISABLED   MODELTYPE      CONTAINERS   AGE
mlserver-1.x                     sklearn        mlserver     10m
my-custom-model-0.x              custom_model   mlserver     10m
ovms-1.x                         openvino_ir    ovms         10m
torchserve-0.x                   pytorch-mar    torchserve   10m
triton-2.x                       keras          triton       10m
```

> ## 4. Deploy the Inference Service according to custom ServingRuntime

Reviewer: ```suggestion
## 4. Deploy the InferenceService using the custom ServingRuntime
```

Author: Done
